### PR TITLE
Finally swap out jackson for circe

### PR DESCRIPTION
**Pull request by alicefuzier**
### What is this PR trying to achieve?
Use only circe as json library (except for the api)
### Who is this change for?
🐱

### CODE REVIEW

File: catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/transformer/source/SierraBibData.scala [hunk 1]
```diff
@@ -3,8 +3,8 @@ package uk.ac.wellcome.transformer.source
 case class SierraBibData(
   id: String,
   title: String,
-  deleted: Boolean,
-  suppressed: Boolean,
+  deleted: Boolean = false,
+  suppressed: Boolean = false,
```
Comment (alicefuzier) [inline catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/transformer/source/SierraBibData.scala]: This is because circe will fail decoding json that doesn't contain those fields unless there is a default value specified

File: catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/transformer/source/SierraItemData.scala [hunk 2]
```diff
@@ -2,7 +2,7 @@ package uk.ac.wellcome.transformer.source
 
 case class SierraItemData(
   id: String,
-  deleted: Boolean,
+  deleted: Boolean = false,
```
Comment (alicefuzier) [inline catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/transformer/source/SierraItemData.scala]: Ditto

File: catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/transformer/transformers/MiroTransformableTransformer.scala [hunk 3]
```diff
@@ -400,35 +400,19 @@ class MiroTransformableTransformer
     * of (key, value) tuples.  Note: we don't use a map because keys aren't
     * guaranteed to be unique.
     */
-  def zipMiroFields(keys: Option[List[String]],
-                    values: Option[List[String]]): List[(String, String)] = {
-    (keys, values) match {
-      case (Some(k), Some(v)) => {
-        if (k.length != v.length) {
+  def zipMiroFields(keys: List[Option[String]],
```
Comment (alicefuzier) [inline catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/transformer/transformers/MiroTransformableTransformer.scala]: This is related to the change on `MiroTransformableData`:`image_award` can contain null values ie `[null, "Biomedical award"]`. While Jackson simply ignores `null`, circe needs the list to explicitly be a `List[Option[String]]` to be able to decode `null`.
Nice side effect is that the logic here becomes much simpler.

File: catalogue_pipeline/transformer/src/test/scala/uk/ac/wellcome/transformer/receive/SQSMessageReceiverTest.scala [hunk 4]
```diff
@@ -125,7 +124,7 @@ class SQSMessageReceiverTest
     val future = recordReceiver.receiveMessage(failingTransformCalmSqsMessage)
 
     whenReady(future.failed) { x =>
-      x shouldBe a[JsonParseException]
+      x shouldBe a[SQSReaderGracefulException]
```
Comment (alicefuzier) [inline catalogue_pipeline/transformer/src/test/scala/uk/ac/wellcome/transformer/receive/SQSMessageReceiverTest.scala]: I don't think the original behaviour here was correct, so fixing it

File: common/src/main/scala/uk/ac/wellcome/models/Item.scala [hunk 5]
```diff
@@ -7,7 +7,6 @@ case class Item(
   sourceIdentifier: SourceIdentifier,
   identifiers: List[SourceIdentifier] = Nil,
   locations: List[Location] = List(),
-  visible: Boolean = true
-) extends Identifiable {
-  @JsonProperty("type") val ontologyType: String = "Item"
-}
+  visible: Boolean = true,
+  @JsonProperty("type") ontologyType: String = "Item"
```
Comment (alicefuzier) [inline common/src/main/scala/uk/ac/wellcome/models/Item.scala]: Had to pull this into the parameter list cause circe relies on shapeless instead of reflection, which doesn't give visibility on members in the class body. 
I did prefer them being constant, but, meh 😐

**Comment by kenoir** [comment]
retrigger travis

File: common/src/main/scala/uk/ac/wellcome/models/License.scala [hunk 6]
```diff
@@ -18,16 +16,20 @@ sealed trait License {
   @JsonProperty("type") val ontologyType: String = "License"
 }
 
-class LicenseDeserialiser extends JsonDeserializer[License] with Logging {
+object License extends Logging {
+  implicit val licenseEncoder = Encoder.instance[License](license => Json.obj(
```
Comment (alicefuzier) [inline common/src/main/scala/uk/ac/wellcome/models/License.scala]: Had to do this again cause shapeless doesn't give visibility on things other than case class parameters. Licenses are case objects so they can't have parameters so, meh 😐

File: common/src/test/scala/uk/ac/wellcome/models/ItemTest.scala [hunk 7]
```diff
@@ -23,13 +24,13 @@ class ItemTest extends FunSpec with Matchers with JsonTestUtil {
       |  "locations": [
       |    {
       |      "locationType": "location",
+      |      "url" : null,
+      |      "credit" : null,
```
Comment (alicefuzier) [inline common/src/test/scala/uk/ac/wellcome/models/ItemTest.scala]: circe is configured to output everything, even nulls. We can configure it differently but I prefer this 'cause it means all fields are always present

File: common/src/test/scala/uk/ac/wellcome/models/ItemTest.scala [hunk 8]
```diff
@@ -23,13 +24,13 @@ class ItemTest extends FunSpec with Matchers with JsonTestUtil {
...
       |      "license": {
       |        "licenseType": "${License_CCBY.licenseType}",
       |        "label": "${License_CCBY.label}",
-      |        "url": "${License_CCBY.url}",
-      |        "type": "License"
-      |      },
-      |      "type": "Location"
+      |        "url": "${License_CCBY.url}"
+      |      }
```
Comment (alicefuzier) [inline common/src/test/scala/uk/ac/wellcome/models/ItemTest.scala]: I only changed `Work` and `Item` to have an `ontlogyType` field that can be serialised by circe. That is because they are `Identifiable` and the id_minter needs the `ontologyType` to save in the database along with the `sourceIdentifier`. 
For other types, I don't actually think we need this. It's more relayed to how we display things in the API, so I would be in favour of removing all of these types out of our internal data model. What do you think?

File: common/src/test/scala/uk/ac/wellcome/models/LicenseTest.scala [hunk 9]
```diff
@@ -1,24 +1,18 @@
...
+import uk.ac.wellcome.circe.jsonUtil._
 
 class LicenseTest extends FunSpec with Matchers {
 
   it("should serialise a License as JSON") {
-    val result = JsonUtil.toJson(License_CCBY)
+    val result = jsonUtil.toJson[License](License_CCBY)
     result.isSuccess shouldBe true
-    result.get shouldBe """{"licenseType":"CC-BY","label":"Attribution 4.0 International (CC BY 4.0)","url":"http://creativecommons.org/licenses/by/4.0/","type":"License"}"""
+    result.get shouldBe """{"licenseType":"CC-BY","label":"Attribution 4.0 International (CC BY 4.0)","url":"http://creativecommons.org/licenses/by/4.0/"}"""
```
Comment (alicefuzier) [inline common/src/test/scala/uk/ac/wellcome/models/LicenseTest.scala]: same as above

File: common/src/main/scala/uk/ac/wellcome/models/Identifiable.scala [hunk 10]
```diff
@@ -8,4 +8,6 @@ trait Identifiable {
   def id: String = canonicalId.getOrElse(
     throw UnidentifiableException()
   )
+  val sourceIdentifier: SourceIdentifier
+  val ontologyType: String
```
Comment (alicefuzier) [inline common/src/main/scala/uk/ac/wellcome/models/Identifiable.scala]: Id minter needs these to be able to mint ids so enforcing them in the base trait

File: catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/transformer/source/MiroTransformableData.scala [hunk 11]
```diff
@@ -1,37 +1,34 @@
...
+  @JsonKey("image_tech_file_size") techFileSize: Option[List[String]],
+  @JsonKey("image_use_restrictions") useRestrictions: Option[String],
+  @JsonKey("image_supp_lettering") suppLettering: Option[String],
+  @JsonKey("image_innopac_id") innopacID: Option[String],
+  @JsonKey("image_credit_line") creditLine: Option[String],
+  @JsonKey("image_source_code") sourceCode: Option[String],
+  @JsonKey("image_library_ref_department") libraryRefDepartment: List[Option[String]] = Nil,
+  @JsonKey("image_library_ref_id") libraryRefId: List[Option[String]] = Nil,
+  @JsonKey("image_award") award: List[Option[String]] = Nil,
+  @JsonKey("image_award_date") awardDate: List[Option[String]] = Nil
```
Comment (alicefuzier) [inline catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/transformer/source/MiroTransformableData.scala]: These have become `List[Option[String]]` instead of `Option[List[String]]`. Check comment on `MiroTransformableTransformer` as to why

File: catalogue_api/api/src/test/scala/uk/ac/wellcome/platform/api/models/DisplayItemTest.scala [hunk 12]
```diff
@@ -63,7 +64,8 @@ class DisplayItemTest extends FunSpec with Matchers {
   }
 
   it("correctly parses an Item without any locations") {
-    val item = JsonUtil.fromJson[Item]("""
+    val item =
+      ApiJsonUtil.fromJson[Item]("""
```
Comment (alexwlchan) [inline catalogue_api/api/src/test/scala/uk/ac/wellcome/platform/api/models/DisplayItemTest.scala]: It's not especially relevant to this patch, but why is there any JSON decoding happening in this test?

File: catalogue_pipeline/id_minter/src/test/scala/uk/ac/wellcome/platform/idminter/IdMinterFeatureTest.scala [hunk 13]
```diff
@@ -88,7 +87,7 @@ class IdMinterFeatureTest
     val miroId = "1234"
     val sqsMessage = generateSqsMessage(miroId)
 
-    sqsClient.sendMessage(idMinterQueue, JsonUtil.toJson(sqsMessage).get)
+    sqsClient.sendMessage(idMinterQueue, toJson(sqsMessage).get)
```
Comment (alexwlchan) [inline catalogue_pipeline/id_minter/src/test/scala/uk/ac/wellcome/platform/idminter/IdMinterFeatureTest.scala]: Should we modify the signature of `sendMessage()` to take a queue and an instance of SQSMessage, and have it do the JSON encoding?

(Again, could be a separate patch!)

File: catalogue_pipeline/id_minter/src/test/scala/uk/ac/wellcome/platform/idminter/steps/IdEmbedderTests.scala [hunk 14]
```diff
@@ -95,8 +95,8 @@ class IdEmbedderTests
         )
     ).thenReturn(Try(throw expectedException))
 
-    val newWorkFuture = idEmbedder.embedId(
-      json = parse(JsonUtil.toJson(originalWork).get).right.get)
+    val newWorkFuture =
+      idEmbedder.embedId(json = parse(toJson(originalWork).get).right.get)
```
Comment (alexwlchan) [inline catalogue_pipeline/id_minter/src/test/scala/uk/ac/wellcome/platform/idminter/steps/IdEmbedderTests.scala]: Sending a work to a JSON string, then unpacking it as a Circe Json object, seems moderately odd.

Maybe we need `toJsonString` and `toCirceJson` in the utils file (or similar)? Depends on how frequently it comes up elsewhere.

File: catalogue_pipeline/ingestor/src/main/scala/uk/ac/wellcome/platform/ingestor/services/WorkIndexer.scala [hunk 15]
```diff
@@ -33,12 +31,11 @@ class WorkIndexer @Inject()(
       "ingestor-index-work",
       () => {
         Future
-          .fromTry(Try {
-            decode[Work](document) match {
-              case Right(work) => work
-              case Left(error) => throw SQSReaderGracefulException(error)
+          .fromTry(
+            fromJson[Work](document).recover {
+              case error: Throwable => throw GracefulFailureException(error)
```
Comment (alexwlchan) [inline catalogue_pipeline/ingestor/src/main/scala/uk/ac/wellcome/platform/ingestor/services/WorkIndexer.scala]: Note to self: this clashes with #1427, and that will need to be rebased atop this later.

File: catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/transformer/source/SierraBibData.scala (hunk 1 above)
Comment (alexwlchan) [reply catalogue_pipeline/transformer/src/main/scala/uk/ac/wellcome/transformer/source/SierraBibData.scala]: When would we not have a value?

File: catalogue_pipeline/transformer/src/test/scala/uk/ac/wellcome/transformer/MiroTransformerFeatureTest.scala [hunk 16]
```diff
@@ -1,6 +1,7 @@
 package uk.ac.wellcome.transformer
 
 import org.scalatest.{FunSpec, Matchers}
+import uk.ac.wellcome.utils.JsonUtil._
```
Comment (alexwlchan) [inline catalogue_pipeline/transformer/src/test/scala/uk/ac/wellcome/transformer/MiroTransformerFeatureTest.scala]: What do these imports add?

File: common/src/main/scala/uk/ac/wellcome/models/Identifiable.scala (hunk 10 above)
Comment (alexwlchan) [reply common/src/main/scala/uk/ac/wellcome/models/Identifiable.scala]: (I’d forgotten these weren’t already there!)

File: common/src/main/scala/uk/ac/wellcome/models/Item.scala (hunk 5 above)
Comment (alexwlchan) [reply common/src/main/scala/uk/ac/wellcome/models/Item.scala]: Yeah, except they aren’t sometimes, so we'd have done this anyway 😕

File: common/src/test/scala/uk/ac/wellcome/models/ItemTest.scala (hunk 8 above)
Comment (alexwlchan) [reply common/src/test/scala/uk/ac/wellcome/models/ItemTest.scala]: As discussed, I think we do need to pass these types around internally.

File: sierra_adapter/sierra_item_merger/src/main/scala/uk/ac/wellcome/platform/sierra_item_merger/services/SierraItemMergerWorkerService.scala [hunk 17]
```diff
@@ -24,12 +25,12 @@ class SierraItemMergerWorkerService @Inject()(
 
   override def processMessage(message: SQSMessage): Future[Unit] =
     // Using Circe here because Jackson creates nulls for empty lists
-    decode[SierraItemRecord](message.body) match {
-      case Right(record) => sierraItemMergerUpdaterService.update(record)
-      case Left(e) =>
+    JsonUtil.fromJson[SierraItemRecord](message.body) match {
```
Comment (alexwlchan) [inline sierra_adapter/sierra_item_merger/src/main/scala/uk/ac/wellcome/platform/sierra_item_merger/services/SierraItemMergerWorkerService.scala]: nit: let's use fromJson for consistency.

File: sierra_adapter/sierra_item_merger/src/main/scala/uk/ac/wellcome/platform/sierra_item_merger/services/SierraItemMergerWorkerService.scala [hunk 18]
```diff
@@ -24,12 +25,12 @@ class SierraItemMergerWorkerService @Inject()(
 
   override def processMessage(message: SQSMessage): Future[Unit] =
     // Using Circe here because Jackson creates nulls for empty lists
```
Comment (alexwlchan) [inline sierra_adapter/sierra_item_merger/src/main/scala/uk/ac/wellcome/platform/sierra_item_merger/services/SierraItemMergerWorkerService.scala]: I think you can lose this comment now!

File: sierra_adapter/sierra_items_to_dynamo/src/main/scala/uk/ac/wellcome/platform/sierra_items_to_dynamo/services/SierraItemsToDynamoWorkerService.scala [hunk 19]
```diff
@@ -21,13 +22,13 @@ class SierraItemsToDynamoWorkerService @Inject()(
 ) extends SQSWorker(reader, system, metrics) {
 
   def processMessage(message: SQSMessage): Future[Unit] =
-    decode[SierraRecord](message.body) match {
-      case Right(record) =>
+    JsonUtil.fromJson[SierraRecord](message.body) match {
```
Comment (alexwlchan) [inline sierra_adapter/sierra_items_to_dynamo/src/main/scala/uk/ac/wellcome/platform/sierra_items_to_dynamo/services/SierraItemsToDynamoWorkerService.scala]: nit: ditto above

File: catalogue_api/api/src/test/scala/uk/ac/wellcome/platform/api/models/DisplayItemTest.scala (hunk 12 above)
Comment (alicefuzier) [reply catalogue_api/api/src/test/scala/uk/ac/wellcome/platform/api/models/DisplayItemTest.scala]: very good question
